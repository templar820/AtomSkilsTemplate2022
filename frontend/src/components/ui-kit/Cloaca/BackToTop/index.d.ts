import * as React from 'react';
import { EmotionStylesType } from 'lib-ui';

export interface BackToTopProps {
    /** send the ref if you need scroll to some element */
    refProp?: React.RefObject<HTMLElement>;
    /** window ref in which the element will scroll */
    windowRefProp?: React.RefObject<HTMLElement>;
    /** body ref which will scroll */
    bodyRefProp?: React.RefObject<HTMLElement>;
    /** Show on scroll  */
    showOnScroll?: boolean;
    /** Canceling of animation  */
    cancelAnimation?: boolean;
    /** Scroll direction */
    direction?: 'toBottom' | 'toTop' | 'toRight' | 'toLeft';
    /** easing */
    easing?: string;
    /** Callback that fires on mouse click */
    onClick?: () => void;
    /** Show at some height or width */
    showAt?: number;
    /** Animation speed */
    animationDuration?: number;
    /** Style */
    style?: EmotionStylesType;
}

declare const BackToTop: React.ComponentType<BackToTopProps>;

export default BackToTop;
