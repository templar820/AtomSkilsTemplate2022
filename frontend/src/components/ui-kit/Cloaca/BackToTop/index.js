import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { css } from '@emotion/core';

import { debounce } from 'lib-root/utils';
import Button from 'lib-ui/Button';
import { setClassName } from 'lib-root/utils/styleMixins';

const ARROW_ROTATE = {
    toBottom: 0,
    toTop: 180,
    toRight: -90,
    toLeft: 90
};

const canUseDOM = !!(typeof window !== 'undefined' && window.document && window.document.createElement);

/**
 * Компонент скроллящий в начало страницы в заданом направлении
 */
class BackToTop extends Component {
    lastScrollTop = 0;

    state = {
        isHidden: true,
        windowBounding: canUseDOM ? document.body.getBoundingClientRect() : {},
        isRendered: false
    };

    getBoundingClientRect = (element, parent) => {
        if (element === document.body) {
            return element.getBoundingClientRect();
        }

        parent = parent || document.body;

        const elementRect = element.getBoundingClientRect();
        const parentRect = parent.getBoundingClientRect();

        return {
            x: elementRect.x - parentRect.x,
            y: elementRect.y - parentRect.y,
            top: elementRect.top - parentRect.top,
            right: elementRect.right - parentRect.right,
            bottom: elementRect.bottom - parentRect.bottom,
            left: elementRect.left - parentRect.left
        };
    };

    scrollFunction = (scrollTo, scrollFactor, axis) => {
        const { easing, animationDuration, cancelAnimation } = this.props;

        let startTime = 0;
        const time = Math.max(0.1, Math.min(Math.abs(scrollFactor - scrollTo) / animationDuration, 0.8));

        const easingEquations = {
            easeOutSine: (pos) => Math.sin(pos * (Math.PI / 2)),
            easeInOutSine: (pos) => -0.5 * (Math.cos(Math.PI * pos) - 1),
            easeInOutQuint: (pos) => {
                const newPos = pos / 0.5;

                if (newPos < 1) {
                    return 0.5 * newPos ** 5;
                }

                return 0.5 * (newPos - 2) ** 5 + 2;
            }
        };

        const tick = () => {
            startTime += 1 / 60;

            const p = startTime / time;
            const t = easingEquations[easing](p);
            const { direction } = this.props;
            const animationPathPoint = cancelAnimation ? scrollTo : scrollFactor + (scrollTo - scrollFactor) * t;

            if (p < 1) {
                this.scrollWindow.requestAnimFrame(tick);

                if (direction === 'toTop' || direction === 'toBottom')
                    this.scrollWindow.scrollTo(axis, animationPathPoint);
                if (direction === 'toRight' || direction === 'toLeft')
                    this.scrollWindow.scrollTo(animationPathPoint, axis);
            } else {
                if (direction === 'toTop' || direction === 'toBottom') this.scrollWindow.scrollTo(axis, scrollTo);
                if (direction === 'toRight' || direction === 'toLeft') this.scrollWindow.scrollTo(scrollTo, axis);
            }
        };

        tick();
    };

    handleVisible = () => {
        const { showAt, direction, showOnScroll } = this.props;
        const { windowBounding } = this.state;

        const scrollStatus = {
            toBottom: Math.abs(windowBounding.top),
            toTop: Math.abs(windowBounding.top),
            toRight: Math.abs(windowBounding.left),
            toLeft: Math.abs(windowBounding.left)
        };

        const isAtRange = scrollStatus[direction] >= showAt;

        const lastScrollDirections = {
            toBottom: showOnScroll && scrollStatus[direction] < this.lastScrollTop,
            toTop: showOnScroll && scrollStatus[direction] > this.lastScrollTop,
            toRight: showOnScroll && scrollStatus[direction] < this.lastScrollTop,
            toLeft: showOnScroll && scrollStatus[direction] > this.lastScrollTop
        };

        this.setState({
            isHidden: !isAtRange || lastScrollDirections[direction]
        });

        this.lastScrollTop = scrollStatus[direction] <= 0 ? 0 : scrollStatus[direction];
    };

    setWindowBounding = () => {
        this.setState({
            windowBounding: canUseDOM ? this.getBoundingClientRect(this.scrollBody, this.scrollWindow) : {}
        });
    };

    scrolling = () => {
        const { direction, refProp } = this.props;

        const { windowBounding } = this.state;
        const { scrollFunction } = this;

        let axis;
        let scrollFactor;
        let scrollTo;

        this.props.onClick && this.props.onClick();

        const isRefScroll = refProp && refProp.current;

        switch (direction) {
            case 'toTop':
                if (isRefScroll) {
                    scrollTo = this.getBoundingClientRect(refProp.current, this.scrollBody).top;
                    scrollFactor = Math.abs(windowBounding.y);
                    axis = Math.abs(windowBounding.x);
                } else {
                    scrollTo = this.getBoundingClientRect(this.scrollBody).top;
                    scrollFactor = Math.abs(windowBounding.y);
                    axis = Math.abs(windowBounding.x);
                }

                break;

            case 'toBottom':
                if (isRefScroll) {
                    scrollTo =
                        this.getBoundingClientRect(refProp.current, this.scrollBody).top + refProp.current.offsetHeight;
                    scrollFactor = Math.abs(windowBounding.y);
                    axis = Math.abs(windowBounding.x);
                } else {
                    scrollTo = this.scrollBody.scrollHeight;
                    scrollFactor = Math.abs(windowBounding.y);
                    axis = Math.abs(windowBounding.x);
                }

                break;
            case 'toRight':
                if (isRefScroll) {
                    scrollTo =
                        this.getBoundingClientRect(refProp.current, this.scrollBody).left + refProp.current.offsetWidth;
                    scrollFactor = Math.abs(windowBounding.x);
                    axis = Math.abs(windowBounding.top);
                } else {
                    scrollTo = this.scrollBody.scrollWidth;
                    scrollFactor = Math.abs(windowBounding.x);
                    axis = Math.abs(windowBounding.top);
                }

                break;
            case 'toLeft':
                if (isRefScroll) {
                    scrollTo = this.getBoundingClientRect(refProp.current, this.scrollBody).left;
                    scrollFactor = Math.abs(windowBounding.x);
                    axis = Math.abs(windowBounding.y);
                } else {
                    scrollTo = this.scrollBody.offsetLeft;
                    scrollFactor = Math.abs(windowBounding.x);
                    axis = Math.abs(windowBounding.top);
                }

                break;
            default:
                break;
        }

        scrollFunction(scrollTo, scrollFactor, axis);
    };

    updateScrollListener = debounce(this.setWindowBounding, 50);
    updateHandleVisible = debounce(this.handleVisible, 50);

    componentDidUpdate = (_, prevState) => {
        if (!prevState.isRendered) {
            if (canUseDOM) {
                this.scrollWindow =
                    this.props.windowRefProp && this.props.windowRefProp.current
                        ? this.props.windowRefProp.current
                        : window;

                this.scrollBody =
                    this.props.bodyRefProp && this.props.bodyRefProp.current
                        ? this.props.bodyRefProp.current
                        : document.body;

                this.scrollWindow.requestAnimFrame = (() => {
                    const clb = (callback) => {
                        window.setTimeout(callback, 1000 / 60);
                    };

                    return (
                        this.scrollWindow.requestAnimationFrame ||
                        this.scrollWindow.webkitRequestAnimationFrame ||
                        this.scrollWindow.mozRequestAnimationFrame ||
                        clb
                    );
                })();

                this.setWindowBounding();
                this.handleVisible();
                this.scrollWindow.addEventListener('scroll', this.updateScrollListener);
                this.scrollWindow.addEventListener('scroll', this.updateHandleVisible);

                this.setState({
                    windowBounding: this.getBoundingClientRect(this.scrollBody, this.scrollWindow)
                });
            }
        }
    };

    componentDidMount = () => {
        this.setState({
            isRendered: true
        });
    };

    componentWillUnmount = () => {
        if (canUseDOM) {
            this.scrollWindow.removeEventListener('scroll', this.updateHandleVisible);
            this.scrollWindow.removeEventListener('scroll', this.updateScrollListener);
        }
    };

    render() {
        const { style, direction, ...rest } = this.props;
        const { isHidden } = this.state;

        if (!canUseDOM) {
            return null;
        }

        return (
            <Button
                className={setClassName({ props: this.props, name: 'back-to-top' })}
                onClick={this.scrolling}
                icon={'Triangle'}
                iconStyle={css`
                    transform: rotate(${ARROW_ROTATE[direction]}deg);
                    width: 40px;
                    min-width: 40px;
                    height: 40px;
                `}
                css={css`
                    opacity: ${isHidden ? 0 : 1};
                    visibility: ${isHidden ? 'hidden' : 'visible'};
                `}
                style={style}
                {...rest}
            />
        );
    }
}

BackToTop.displayName = 'BackToTop';
BackToTop.propTypes = {
    /** send the ref if you need scroll to some element */
    refProp: PropTypes.shape({ current: PropTypes.instanceOf(Element) }),
    /** window ref in which the element will scroll */
    windowRefProp: PropTypes.shape({ current: PropTypes.instanceOf(Element) }),
    /** body ref which will scroll */
    bodyRefProp: PropTypes.shape({ current: PropTypes.instanceOf(Element) }),
    /** Show on scroll  */
    showOnScroll: PropTypes.bool,
    /** Canceling of animation  */
    cancelAnimation: PropTypes.bool,
    /** Scroll direction */
    direction: PropTypes.oneOf(['toBottom', 'toTop', 'toRight', 'toLeft']),
    /** easing */
    easing: PropTypes.string,
    /** Callback that fires on mouse click */
    onClick: PropTypes.func,
    /** Show at some height or width */
    showAt: PropTypes.number,
    /** Animation speed */
    animationDuration: PropTypes.number,
    /** Style */
    style: PropTypes.oneOfType([PropTypes.func, PropTypes.object, PropTypes.string])
};

BackToTop.defaultProps = {
    refProp: undefined,
    direction: 'toTop',
    cancelAnimation: false,
    showOnScroll: false,
    easing: 'easeInOutSine',
    showAt: 1000,
    animationDuration: 2000,
    style: {
        position: 'fixed',
        left: 50,
        bottom: 50
    }
};

export default BackToTop;
