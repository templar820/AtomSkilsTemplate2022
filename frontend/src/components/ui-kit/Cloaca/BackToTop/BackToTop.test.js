import React from 'react';
import ReactDOM from 'react-dom';
import serializer from 'jest-emotion';
import Enzyme, { mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import toJson from 'enzyme-to-json';

import BackToTop from './';

expect.addSnapshotSerializer(serializer);

Enzyme.configure({ adapter: new Adapter() });

const myMockFn = jest.fn();

const props = {
    onClick: () => myMockFn('onClick'),
    cancelAnimation: false,
    direction: 'toBottom',
    animationDuration: 2000,
    scrollTo: 0,
    easing: 'easeOutSine',
    showAt: 0,
    showOnScroll: false,
    style: {
        position: 'fixed',
        left: 10,
        top: 100
    }
};

describe('<BackToTop />', () => {
    let wrapper;

    beforeEach(() => {
        wrapper = mount(<BackToTop {...props} />);
    });

    afterEach(() => {
        wrapper.unmount();
    });

    test('should match snapshot', () => {
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    test('renders without crashing', () => {
        const div = document.createElement('div');
        ReactDOM.render(<BackToTop />, div);
        ReactDOM.unmountComponentAtNode(div);
    });

    test('should render correctly', () => {
        expect(wrapper.find('button').length).toBe(1);
        expect(wrapper.props()).toHaveProperty('style', props.style);
    });

    test('should call onClick function', () => {
        wrapper.simulate('click');
        expect(myMockFn).toHaveBeenCalledWith('onClick');
    });
});
