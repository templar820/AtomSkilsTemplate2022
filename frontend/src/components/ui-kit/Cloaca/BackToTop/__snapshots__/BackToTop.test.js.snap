// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BackToTop /> should match snapshot 1`] = `
.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.emotion-0 {
  fill: currentColor;
  width: 32px;
  min-width: 32px;
  height: 32px;
  -webkit-transform: rotate(0deg);
  -ms-transform: rotate(0deg);
  transform: rotate(0deg);
  width: 40px;
  min-width: 40px;
  height: 40px;
}

.emotion-4 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  min-width: 24px;
  width: 60px;
  height: 60px;
  margin: 0;
  padding: 0;
  border: 1px solid;
  border-radius: 100%;
  font-weight: 500;
  font-size: 15px;
  line-height: 1.6;
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  -webkit-transition: 0.3s all;
  transition: 0.3s all;
  color: #ffffff;
  white-space: nowrap;
  background-color: #00B956;
  border-color: #00B956;
  opacity: 1;
  visibility: visible;
}

.emotion-4.hover,
.emotion-4:hover {
  background-color: #10E272;
  border-color: #10E272;
  color: #ffffff;
}

.emotion-4:focus {
  outline: 0;
}

.emotion-4:active,
.emotion-4.active {
  border-color: transparent;
  background-color: #404D46;
  box-shadow: inset 0 1px 3px #333333;
  color: #ffffff;
}

.emotion-4:disabled {
  background-color: #D8D8D8;
  border-color: #D8D8D8;
  cursor: not-allowed;
  color: #ffffff;
}

.emotion-4:after {
  content: '';
  min-height: inherit;
  font-size: 0;
}

<BackToTop
  animationDuration={2000}
  cancelAnimation={false}
  direction="toBottom"
  easing="easeOutSine"
  onClick={[Function]}
  scrollTo={0}
  showAt={0}
  showOnScroll={false}
  style={
    Object {
      "left": 10,
      "position": "fixed",
      "top": 100,
    }
  }
>
  <Button
    animationDuration={2000}
    cancelAnimation={false}
    className="CLR-back-to-top"
    colorScheme="primary"
    css={
      Object {
        "map": undefined,
        "name": "1hk0kxi-BackToTop",
        "next": undefined,
        "styles": "opacity:1;visibility:visible;;label:BackToTop;",
        "toString": [Function],
      }
    }
    disabled={false}
    easing="easeOutSine"
    icon="Triangle"
    iconStyle={
      Object {
        "map": undefined,
        "name": "1cak18s-BackToTop",
        "next": undefined,
        "styles": "transform:rotate(0deg);width:40px;min-width:40px;height:40px;;label:BackToTop;",
        "toString": [Function],
      }
    }
    isLoading={false}
    onClick={[Function]}
    scrollTo={0}
    showAt={0}
    showOnScroll={false}
    size="lg"
    style={
      Object {
        "left": 10,
        "position": "fixed",
        "top": 100,
      }
    }
    type="button"
  >
    <StyledButton
      animationDuration={2000}
      cancelAnimation={false}
      className="CLR-back-to-top CLR-button"
      colorScheme="primary"
      colors={
        Object {
          "GrayScale_0": "#ffffff",
          "GrayScale_100": "#EDEDED",
          "GrayScale_150": "#E5E5E5",
          "GrayScale_200": "#D8D8D8",
          "GrayScale_50": "#fafafa",
          "GrayScale_500": "#999999",
          "GrayScale_55": "#f6f6f6",
          "GrayScale_600": "#7C7C7C",
          "GrayScale_700": "#333333",
          "GrayScale_900": "#000000",
          "error": "#F62434",
          "errorAccent": "#EB5A40",
          "info": "#34AAF2",
          "infoAccent": "#5BD9E5",
          "primary": "#00B956",
          "primaryAccent": "#10E272",
          "secondary": "#731982",
          "secondaryAccent": "#A500BF",
          "tertiary": "#444189",
          "tertiaryAccent": "#404D46",
          "warning": "#FFA717",
          "warningAccent": "#F36F20",
        }
      }
      css={
        Object {
          "map": undefined,
          "name": "1hk0kxi-BackToTop",
          "next": undefined,
          "styles": "opacity:1;visibility:visible;;label:BackToTop;",
          "toString": [Function],
        }
      }
      disabled={false}
      easing="easeOutSine"
      icon="Triangle"
      iconStyle={
        Object {
          "map": undefined,
          "name": "1cak18s-BackToTop",
          "next": undefined,
          "styles": "transform:rotate(0deg);width:40px;min-width:40px;height:40px;;label:BackToTop;",
          "toString": [Function],
        }
      }
      isLoading={false}
      onClick={[Function]}
      scrollTo={0}
      showAt={0}
      showOnScroll={false}
      size="lg"
      style={
        Object {
          "left": 10,
          "position": "fixed",
          "top": 100,
        }
      }
      type="button"
    >
      <button
        className="CLR-back-to-top CLR-button emotion-4 emotion-5"
        disabled={false}
        onClick={[Function]}
        size="lg"
        style={
          Object {
            "left": 10,
            "position": "fixed",
            "top": 100,
          }
        }
        type="button"
      >
        <StyledButtonInner>
          <div
            className="emotion-2 emotion-3"
          >
            <withColors(InlineIcons)
              color="currentColor"
              css={
                Object {
                  "map": undefined,
                  "name": "1cak18s-BackToTop",
                  "next": undefined,
                  "styles": "transform:rotate(0deg);width:40px;min-width:40px;height:40px;;label:BackToTop;",
                  "toString": [Function],
                }
              }
              disabled={false}
              h="32px"
              height={null}
              icon="Triangle"
              w="32px"
              width={null}
            >
              <InlineIcons
                color="currentColor"
                colors={
                  Object {
                    "GrayScale_0": "#ffffff",
                    "GrayScale_100": "#EDEDED",
                    "GrayScale_150": "#E5E5E5",
                    "GrayScale_200": "#D8D8D8",
                    "GrayScale_50": "#fafafa",
                    "GrayScale_500": "#999999",
                    "GrayScale_55": "#f6f6f6",
                    "GrayScale_600": "#7C7C7C",
                    "GrayScale_700": "#333333",
                    "GrayScale_900": "#000000",
                    "error": "#F62434",
                    "errorAccent": "#EB5A40",
                    "info": "#34AAF2",
                    "infoAccent": "#5BD9E5",
                    "primary": "#00B956",
                    "primaryAccent": "#10E272",
                    "secondary": "#731982",
                    "secondaryAccent": "#A500BF",
                    "tertiary": "#444189",
                    "tertiaryAccent": "#404D46",
                    "warning": "#FFA717",
                    "warningAccent": "#F36F20",
                  }
                }
                css={
                  Object {
                    "map": undefined,
                    "name": "1cak18s-BackToTop",
                    "next": undefined,
                    "styles": "transform:rotate(0deg);width:40px;min-width:40px;height:40px;;label:BackToTop;",
                    "toString": [Function],
                  }
                }
                disabled={false}
                h="32px"
                height={null}
                icon="Triangle"
                w="32px"
                width={null}
              >
                <Styled(svg)
                  className="CLR-inline-icons"
                  color="currentColor"
                  colors={
                    Object {
                      "GrayScale_0": "#ffffff",
                      "GrayScale_100": "#EDEDED",
                      "GrayScale_150": "#E5E5E5",
                      "GrayScale_200": "#D8D8D8",
                      "GrayScale_50": "#fafafa",
                      "GrayScale_500": "#999999",
                      "GrayScale_55": "#f6f6f6",
                      "GrayScale_600": "#7C7C7C",
                      "GrayScale_700": "#333333",
                      "GrayScale_900": "#000000",
                      "error": "#F62434",
                      "errorAccent": "#EB5A40",
                      "info": "#34AAF2",
                      "infoAccent": "#5BD9E5",
                      "primary": "#00B956",
                      "primaryAccent": "#10E272",
                      "secondary": "#731982",
                      "secondaryAccent": "#A500BF",
                      "tertiary": "#444189",
                      "tertiaryAccent": "#404D46",
                      "warning": "#FFA717",
                      "warningAccent": "#F36F20",
                    }
                  }
                  css={
                    Object {
                      "map": undefined,
                      "name": "1cak18s-BackToTop",
                      "next": undefined,
                      "styles": "transform:rotate(0deg);width:40px;min-width:40px;height:40px;;label:BackToTop;",
                      "toString": [Function],
                    }
                  }
                  disabled={false}
                  height="32px"
                  width="32px"
                >
                  <svg
                    className="CLR-inline-icons emotion-0 emotion-1"
                    color="currentColor"
                    disabled={false}
                    height="32px"
                    width="32px"
                  >
                    <use
                      xlinkHref="#Triangle"
                    />
                  </svg>
                </Styled(svg)>
              </InlineIcons>
            </withColors(InlineIcons)>
          </div>
        </StyledButtonInner>
      </button>
    </StyledButton>
  </Button>
</BackToTop>
`;
